### PR TITLE
Adds methods for checking repairability

### LINKED ISSUE #5622: Make methods in Item for check if a Material can repair a Item
### Is your feature request related to a problem?

I making changes in the Anvil Event, for enchantments in Items customs, i want a easy form to check if the material in the right slot can repair the left slot.

### Describe the solution you'd like.

Make a method like ItemStack::canRepair(Material material) return boolean if the material can repair the item (not sure if this method can be better in Material or ItemStack).

### Describe alternatives you've considered.

- Make a switch to set what materials can repair a item
- Using nms you can use a method what does this task.
`
ItemStack item = new ItemStack(Material.DIAMOND_SWORD);
ItemStack itemRepair = new ItemStack(Material.DIAMOND);
CraftItemStack.asNMSCopy(item).getItem().a(CraftItemStack.asNMSCopy(item),CraftItemStack.asNMSCopy(itemRepair));`
- 

### Agreements

- [X] I have searched for and ensured there isn't already an open issue regarding this.
- [X] I have ensured the feature I'm requesting isn't already in the latest supported Paper build.

### Other

_No response_

### PR DESCRIPTION
**Pull request by Machine-Maker**
Closes #5622

### CODE REVIEW

File: Spigot-API-Patches/0300-ItemStack-repair-check-API.patch [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
++    /**
++     * Checks if an itemstack can be repaired with another itemstack.
++     * <p>
++     * <b>Both Itemstacks' materials must be items ({@link Material#isItem()}).</b>
++     *
++     * @param itemToBeRepaired the itemstack to be repaired, must be item
++     * @param repairMaterial the repair material, must be item
++     * @return true if valid repair, false if not
++     */
++    public boolean isValidRepairItemStack(ItemStack itemToBeRepaired, ItemStack repairMaterial);
```
Comment (Proximyst) [inline Spigot-API-Patches/0300-ItemStack-repair-check-API.patch]: Nullability annotations

File: Spigot-Server-Patches/0730-ItemStack-repair-check-API.patch [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
++++ b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+@@ -487,6 +487,14 @@ public final class CraftMagicNumbers implements UnsafeValues {
+         return io.papermc.paper.inventory.ItemRarity.values()[getItem(itemStack.getType()).getItemStackRarity(CraftItemStack.asNMSCopy(itemStack)).ordinal()];
+     }
+ 
++    @Override
++    public boolean isValidRepairItemStack(org.bukkit.inventory.ItemStack itemToBeRepaired, org.bukkit.inventory.ItemStack repairMaterial) {
++        if (!itemToBeRepaired.getType().isItem() || !repairMaterial.getType().isItem()) {
++            throw new IllegalArgumentException("Both itemstacks must have a type that is an item!");
++        }
```
Comment (Proximyst) [inline Spigot-Server-Patches/0730-ItemStack-repair-check-API.patch]: Guava Preconditions

File: Spigot-API-Patches/0300-ItemStack-repair-check-API.patch [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+--- a/src/main/java/org/bukkit/UnsafeValues.java
++++ b/src/main/java/org/bukkit/UnsafeValues.java
+@@ -155,6 +155,17 @@ public interface UnsafeValues {
+      */
+     public io.papermc.paper.inventory.ItemRarity getItemStackRarity(ItemStack itemStack);
+ 
++    /**
++     * Checks if an itemstack can be repaired with another itemstack.
++     * <p>
++     * <b>Both Itemstacks' materials must be items ({@link Material#isItem()}).</b>
```
Comment (Proximyst) [inline Spigot-API-Patches/0300-ItemStack-repair-check-API.patch]: Why? I don't see a reason for this, as blocks obviously cannot be repaired, so should rather always return `false`.

**Review by Proximyst — CHANGES_REQUESTED** [review]

File: Spigot-API-Patches/0300-ItemStack-repair-check-API.patch (hunk 3 above)
Comment (Machine-Maker) [reply Spigot-API-Patches/0300-ItemStack-repair-check-API.patch]: Yeah, ok. removes the need for the preconditions, can just return false

**Review by Proximyst — APPROVED** [review]

**Review by kennytv — APPROVED** [review]
Thonk

**Pull request merged**